river/Cursor.zig

```diff
@@ -802,26 +935,92 @@ fn processMotion(self: *Self, device: *wlr.InputDevice, time: u32, delta_x: f64,
             data.delta_y = dy - @trunc(dy);
 
             const border_width = if (data.view.draw_borders) server.config.border_width else 0;
-
-            // Set width/height of view, clamp to view size constraints and output dimensions
             const box = &data.view.pending.box;
-            box.width = @intCast(u32, math.max(0, @intCast(i32, box.width) + @floatToInt(i32, dx)));
-            box.height = @intCast(u32, math.max(0, @intCast(i32, box.height) + @floatToInt(i32, dy)));
+            const constraints = data.view.getConstraints();
+
+            // Avoid potential integer overflow. No reasonable client will have
+            // a minimum size even remotely as large, so this should not be
+            // interfering with well behaved toplevels, but it prevents bad view
+            // constraints from crashing us. A max size larger than this is
+            // pretty common however as most views do not have a max size, which
+            // means they have max_width = max_height = maxInt(u32).
+            const max = math.maxInt(i32) / 2;
+            const min_width = @intCast(i32, math.max(View.min_size, math.min(max, constraints.min_width)));
+            const min_height = @intCast(i32, math.max(View.min_size, math.min(max, constraints.min_height)));
+            const max_width = @intCast(i32, math.min(max, constraints.max_width));
+            const max_height = @intCast(i32, math.min(max, constraints.max_height));
+
+            // Using two points instead of a box makes this logic a bit more
+            // readable for humans. The compiler should optimize the conversion out.
+            var botright_x = box.x + @intCast(i32, box.width);
+            var botright_y = box.y + @intCast(i32, box.height);
+
+            if (data.resize_direction == .top or
+                data.resize_direction == .topleft or
+                data.resize_direction == .topright)
+            {
+                const min_y = math.max(border_width, botright_y - max_height);
+                const max_y = botright_y - min_height;
+                box.y = math.clamp(box.y + @floatToInt(i32, dy), min_y, max_y);
+            } else if (data.resize_direction == .bottom or
+                data.resize_direction == .botleft or
+                data.resize_direction == .botright)
+            {
+                const min_y = box.y + min_height;
+                const max_y = box.y + max_height;
+                botright_y = math.clamp(botright_y + @floatToInt(i32, dy), min_y, max_y);
+            }
```

Comment:
I think this would be cleaner as a switch instead of an if chain:
```zig
switch (data.resize_direction) {
   .top, .topleft, .topright => { ... },
   .bottom, .botleft, .botright => { ... },
   .left, .right =>  {}, 
};
```
same below for the x axis